leftwm-core/src/event_loop.rs

```diff
@@ -1,123 +1,194 @@
 use crate::{child_process::Nanny, config::Config};
-use crate::{CommandPipe, DisplayServer, Manager, Mode, StateSocket, Window};
+use crate::{
+    Command, CommandPipe, DisplayEvent, DisplayServer, Manager, Mode, StateSocket, Window,
+};
 use std::path::{Path, PathBuf};
-use std::sync::{atomic::Ordering, Once};
+use std::sync::atomic::Ordering;
+
+/// Errors which can appear while running the event loop.
+#[derive(thiserror::Error, Debug, Clone, PartialEq, Eq, Hash)]
+pub enum Error {
+    #[error("Couldn't create the file: '{0}'")]
+    CreateFile(String),
+
+    #[error("Couldn't connect to file: '{0}'")]
+    ConnectToFile(String),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum EventResponse {
+    None,
+    DisplayRefreshNeeded,
+}
 
 impl<C: Config, SERVER: DisplayServer> Manager<C, SERVER> {
-    /// # Panics
-    /// This function panics if it can't create or write to the command file.
-    pub async fn event_loop(mut self) {
-        let socket_file = place_runtime_file("current_state.sock")
-            .expect("ERROR: couldn't create current_state.sock");
-        let mut state_socket = StateSocket::default();
-        state_socket
-            .listen(socket_file)
-            .await
-            .expect("ERROR: couldn't connect to current_state.sock");
-
-        let file_name = CommandPipe::pipe_name();
-        let pipe_file = place_runtime_file(&file_name)
-            .unwrap_or_else(|_| panic!("ERROR: couldn't create {}", file_name.display()));
-        let mut command_pipe = CommandPipe::new(pipe_file)
-            .await
-            .unwrap_or_else(|_| panic!("ERROR: couldn't connect to {}", file_name.display()));
-
-        // Start the current theme.
-        let after_first_loop: Once = Once::new();
+    /// Starts the event loop of leftwm
+    ///
+    /// # Errors
+    /// `EventResponse` if the initialisation of the command pipe or/and the state socket failed.
+    pub async fn event_loop(mut self) -> Result<(), Error> {
+        let mut state_socket = get_state_socket().await?;
+        let mut command_pipe = get_command_pipe().await?;
+
+        self.call_up_scripts();
 
         // Main event loop.
-        let mut event_buffer = vec![];
-        loop {
-            if self.state.mode == Mode::Normal {
-                state_socket.write_manager_state(&self.state).await.ok();
-            }
-            self.display_server.flush();
+        let mut event_buffer: Vec<DisplayEvent> = vec![];
+        while self.should_keep_running(&mut state_socket).await {
+            self.update_manager_state(&mut state_socket).await;
+            self.display_server.flush(); // is that needed?
```

Comment:
Seems to be needed (i found with lefthk that it seemed to make a difference, atleast on startup).

Reply:
> i found with lefthk that it seemed to make a difference, atleast on startup

may I ask what kind of difference it does?

Reply:
Xlib events not coming in, i suppose sometimes they may get blocked up and needs flushing through to get them going again

Reply:
alright, I removed the comment